### PR TITLE
chore: add test server stub

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright/src/runner/testServer.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  const config = await loadConfigFromFile(configFile);
+  if (!config)
+    return;
+
+  const dispatcher = new Dispatcher(config);
+  const wss = new WSServer({
+    onConnection(request: http.IncomingMessage, url: URL, ws: WebSocket, id: string) {
+      ws.on('message', async message => {
+        const { id, method, params } = JSON.parse(message.toString());
+        const result = await (dispatcher as any)[method](params);
```
Comment (dgozman) [inline packages/playwright/src/runner/testServer.ts:52]: try/catch and send the error response? Or at least not crash :)

File: packages/playwright/src/runner/testServer.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+class Dispatcher {
+  private _config: FullConfigInternal;
+
+  constructor(config: FullConfigInternal) {
+    this._config = config;
+  }
+
+  async test(params: { mode: 'list' | 'run', locations: string[], options: PlaywrightTestOptions, reporter: string, env: NodeJS.ProcessEnv }) {
+    for (const name in params.env)
+      process.env[name] = params.env[name];
```
Comment (dgozman) [inline packages/playwright/src/runner/testServer.ts:74]: Do we want to undo the `env` changes at the end of the method?

File: packages/playwright/src/runner/testServer.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    if (params.mode === 'list')
+      await listTests(this._config, params.reporter, params.locations);
+    if (params.mode === 'run')
+      await runTests(this._config, params.reporter, params.locations, params.options);
+  }
+}
+
+
+async function listTests(config: FullConfigInternal, reporterPath: string, locations: string[] | undefined) {
+  config.cliArgs = [...(locations || []), '--reporter=null'];
```
Comment (dgozman) [inline packages/playwright/src/runner/testServer.ts:84]: I am not really sure `--reporter=null` will be parsed at this point.

File: packages/playwright/src/runner/testServer.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  config.configCLIOverrides.preserveOutputDir = true;
+  config.configCLIOverrides.use =  {
+    trace: options.trace,
+    headless: options.headed ? false : undefined,
+    _optionContextReuseMode: options.reuseContext ? 'when-possible' : undefined,
+    _optionConnectOptions: options.connectWsEndpoint ? { wsEndpoint: options.connectWsEndpoint } : undefined,
+  };
+
+  const reporter = new InternalReporter(new Multiplexer(await createReporters(config, 'run')));
+  const taskRunner = createTaskRunnerForWatch(config, reporter);
```
Comment (dgozman) [inline packages/playwright/src/runner/testServer.ts:118]: This does not include global setup tasks. Did you mean to run them at the start of the dispatcher?

File: packages/playwright/src/runner/testServer.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  const testRun = new TestRun(config, reporter);
+  reporter.onConfigure(config.config);
+  const stop = new ManualPromise();
```
Comment (dgozman) [inline packages/playwright/src/runner/testServer.ts:121]: This one is unused, remove?

**Review by dgozman — APPROVED** [review]

**Comment by pavelfeldman** [comment]
This change was not about the test server - test server was there to show you why the other changes were done. I'll follow-up with the server rewrite anyways.

**Pull request merged**